### PR TITLE
this is a fake PR just for the sake of initial code review

### PR DESCRIPTION
**Pull request by yuvalif**
please consider that the opened issues try to cover some of the problems with the code
please look into documentation added here: https://github.com/nmstate/kubernetes-nmstate/pull/15
for info on how the product SHOULD behave

### CODE REVIEW

File: cmd/client/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    yum clean all
+
+RUN pip install -U \
+      pip \
+      setuptools \
+      pbr \
+      tox \
+      pluggy==0.6 \
+      PyGObject \
+      pytest-cov \
```
Comment (phoracek) [inline cmd/client/Dockerfile:14]: I don't think you need all of these dependencies. testing libraries, tox and gcc are probably not needed.

You could make one shared image and build client and state-controller. Or even better, nmstate could ship their base image. They know the best what dependencies are needed to run it.

Just an idea, no need to change it.

File: cmd/client/main.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+	nmstateioclientset "github.com/nmstate/k8s-node-net-conf/pkg/client/clientset/versioned"
+	"github.com/nmstate/k8s-node-net-conf/pkg/nmstatectl"
+	"github.com/nmstate/k8s-node-net-conf/pkg/utils"
+)
+
+var (
+	kubeconfig = flag.String("kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	master     = flag.String("master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	namespace  = flag.String("n", "", "The namespace where the CRDs are created. If left blank and running via pod, it will be taken from there.")
```
Comment (phoracek) [inline cmd/client/main.go:19]: CRDs should be cluster resources, not namespaced https://github.com/nmstate/kubernetes-nmstate/issues/17.

File: cmd/client/main.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	crdType    = flag.String("type", "state", "state|policy. Whether client should handle 'state' or 'policy' CRDs.")
+	hostname   = flag.String("host", "", "Name of the host on which to enforce and report state. If left blank and running via pod, it will be taken from there.")
+)
+
+func main() {
```
Comment (phoracek) [inline cmd/client/main.go:24]: This looks like a one-shot service. How do I trigger node reconfiguration?

File: cmd/policy-controller/controller.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+	clientset "github.com/nmstate/k8s-node-net-conf/pkg/client/clientset/versioned"
+	informers "github.com/nmstate/k8s-node-net-conf/pkg/client/informers/externalversions/nmstate.io/v1"
+	listers "github.com/nmstate/k8s-node-net-conf/pkg/client/listers/nmstate.io/v1"
+)
+
+// Controller is the controller implementation for NodeNetConfPolicy resources
+type Controller struct {
+	// kubeclientset is a standard kubernetes clientset
+	kubeclientset kubernetes.Interface
```
Comment (phoracek) [inline cmd/policy-controller/controller.go:18]: Cannot you use the patched one also for standard resources?

File: cmd/policy-controller/controller.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	// Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// NewController returns a new state controller
+func NewController(
+	kubeclientset kubernetes.Interface,
+	nmstateclientset clientset.Interface,
+	policyInformer informers.NodeNetConfPolicyInformer) *Controller {
+	return nil
```
Comment (phoracek) [inline cmd/policy-controller/controller.go:41]: why does it take any attributes if it returns nil anyway?

File: cmd/policy-controller/main.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+func main() {
+	flag.Parse()
+
+	// set up signals so we handle the first shutdown signal gracefully
+	stopCh := signals.SetupSignalHandler()
+
+	cfg, err := clientcmd.BuildConfigFromFlags(masterURL, kubeconfig)
+	if err != nil {
```
Comment (phoracek) [inline cmd/policy-controller/main.go:37]: You could save some lines defining something like this:

```go
func FailIf(err error, msg string):
    if err != nil {
        klog.Fatalf(msg, err.Error())
    }
```

File: cmd/policy-controller/main.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	controller := NewController(kubeClient, nmstateClient,
+		nmstateInformerFactory.Nmstate().V1().NodeNetConfPolicies(),
+	)
+
+	// notice that there is no need to run Start methods in a separate goroutine. (i.e. go kubeInformerFactory.Start(stopCh)
+	// Start method is non-blocking and runs all registered informers in a dedicated goroutine.
+	kubeInformerFactory.Start(stopCh)
+	nmstateInformerFactory.Start(stopCh)
+
+	if err = controller.Run(2, stopCh); err != nil {
```
Comment (phoracek) [inline cmd/policy-controller/main.go:63]: what does the `2` stand for? could you use a constant for it?

File: cmd/state-controller/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+      pytest-cov \
+      pytest==3.5.1
+
+RUN git clone https://github.com/nmstate/nmstate.git && \
+    cd nmstate && \
+    pip install --upgrade .
+
+COPY ./state-controller /
+
+ENTRYPOINT ["/state-controller"]
```
Comment (phoracek) [inline cmd/state-controller/Dockerfile:23]: Not sure how klog logging works, but with glog, I usually need to raise verbosity and set it to log to stderr.

**Review by phoracek — CHANGES_REQUESTED** [review]

File: cmd/client/Dockerfile (hunk 1 above)
Comment (yuvalif) [reply cmd/client/Dockerfile:14]: agree. please see here: https://github.com/nmstate/kubernetes-nmstate/issues/9

File: cmd/client/main.go (hunk 3 above)
Comment (yuvalif) [reply cmd/client/main.go:24]: the client is intendend to be a one shot, to be used by an external management system. the controller the a daemon type of deployment

File: cmd/policy-controller/controller.go (hunk 4 above)
Comment (yuvalif) [reply cmd/policy-controller/controller.go:18]: what do you mean by "standard"?

File: cmd/policy-controller/controller.go (hunk 5 above)
Comment (yuvalif) [reply cmd/policy-controller/controller.go:41]: this is not implemented yet. only state controller implemented

File: cmd/policy-controller/main.go (hunk 7 above)
Comment (yuvalif) [reply cmd/policy-controller/main.go:63]: it is for the number of threads. will probably just remove it: https://github.com/nmstate/kubernetes-nmstate/issues/9

File: cmd/state-controller/Dockerfile (hunk 8 above)
Comment (yuvalif) [reply cmd/state-controller/Dockerfile:23]: klog actually didn't work for me. just used fmt... anyway: https://github.com/nmstate/kubernetes-nmstate/issues/4

**Comment by yuvalif** [comment]
not needed anymore

**Pull request closed without merging**